### PR TITLE
In-memory Engine: Reduce `HybridEngine` scope

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/17181

What's Changed:

```commit-message
In TiKV, the `KvEngine` trait provides the write path abstraction
(`WriteBatch`) and read path abstraction (`Snapshot`). The
`HybridEngine` implements the `KvEngine` trait, allowing it to be
integrated into TiKV.

However, as described in issue #17181, there is a major downside: all
components and features in TiKV use the `HybridEngine`, which can
introduce unexpected compatibility issues (e.g., BR restore failures
with checksum errors). This is not worthwhile because the `HybridEngine`
is designed specifically for accelerating coprocessor read speed in
scenarios with many MVCC workloads.

This PR shrinks the `HybridEngine` scope from being globally enabled to
only being enabled in the coprocessor. The changes include:
- Adding a new observer, `WriteBatchObserver`, to observe write requests
  via raftstore coprocessor hooks.
- Adding a new observer, `SnapshotObserver`, to observe snapshot (and
  `read_ts`) via raftstore coprocessor hooks.
- Adding a new method to the `RaftKv` struct called `async_in_memory_snapshot`,
  which takes snapshots from the `HybridEngine`.

This PR also introduces an architectural change, transforming the
in-memory engine from functioning as an `KvEngine` to acting as an
MVCC-aware cache. Conceptually, the in-memory engine is shifted from
the storage engine layer to the MVCC layer. This adjustment better
aligns with its purpose of improving MVCC performance by caching recent
versions, rather than serving as a full-fledged storage engine.

    ┌──────────────┐      ┌──────────────┐
    │ Copr/Storage │      │ Copr/Storage │
    ├──────────────┤      ├──────────────┤
    │    MVCC      │      │     MVCC     │
    ├──────────────┤      │ ┌──────────┐ │
    │ Raftstore    │ ───> │ │   IME    │ │
    ├──────────────┤      ├─┴──────────┴─┤
    │    IME       │      │  Raftstore   │
    │┌────────────┐│      ├──────────────┤
    ││  RocksDB   ││      │    RocksDB   │
    └┴────────────┴┘      └──────────────┘

Additionally, this PR brings two unexpected benefits*:
- Reduced TiKV binary size: From 1.7GB to 1.2GB.
- Faster build time: From 32 minutes to 20 minutes.

*: Results were obtained using the `make release` command.
```

### Check List

Tests 

- Unit test
- Integration test
- [ ] Manual test (add detailed scripts or steps below)

### Release note

```release-note
None
```

### CODE REVIEW

File: components/hybrid_engine/src/engine.rs [hunk 1]
```diff
@@ -97,43 +131,14 @@ where
...
-                        .inc();
-                    None
-                }
-            }
+            new_in_memory_snapshot(&self.range_cache_engine, ctx, disk_snap.sequence_number())
         } else {
             RANGE_CACHEN_SNAPSHOT_ACQUIRE_FAILED_REASON_COUNT_STAIC
                 .no_read_ts
                 .inc();
+            SNAPSHOT_TYPE_COUNT_STATIC.rocksdb.inc();
```
Comment (SpadeA-Tang) [inline components/hybrid_engine/src/engine.rs:139]: why this change

File: components/raftstore/src/store/worker/read.rs [hunk 2]
```diff
@@ -138,10 +142,16 @@ pub trait ReadExecutor {
                     }
                 }
                 CmdType::Snap => {
-                    let snapshot = RegionSnapshot::from_snapshot(
+                    let mut snapshot = RegionSnapshot::from_snapshot(
                         self.get_snapshot(snap_ctx.clone(), &local_read_ctx),
                         region.clone(),
                     );
+                    if let Some(snap_ctx) = snap_ctx.as_ref() {
```
Comment (SpadeA-Tang) [inline components/raftstore/src/store/worker/read.rs:149]: Please add some comments for this.

File: components/raftstore/src/store/region_snapshot.rs [hunk 3]
```diff
@@ -74,6 +91,40 @@ where
...
             bucket_meta: None,
+            observed_snap: None,
+        }
+    }
+
+    pub fn set_observed_snapshot(&mut self, observed_snap: Box<dyn ObservedSnapshot>) {
+        self.observed_snap = Some(Arc::new(Mutex::new(Some(observed_snap))));
+    }
+
+    /// Replace underlying snapshot with its observed snapshot.
```
Comment (SpadeA-Tang) [inline components/raftstore/src/store/region_snapshot.rs:102]: Maybe better to add reason/senarios why we need this replacement.

File: components/server/src/common.rs [hunk 4]
```diff
@@ -695,48 +695,24 @@ impl<T: fmt::Display + Send + 'static> Stop for LazyWorker<T> {
...
-        memory_engine.set_disk_engine(disk_engine.clone());
-        if let Some(pd_client) = pd_client.as_ref() {
-            memory_engine.start_hint_service(
-                <RangeCacheMemoryEngine as RangeCacheEngine>::RangeHintService::from(
-                    pd_client.clone(),
-                ),
-            )
-        }
-        HybridEngine::new(disk_engine, memory_engine)
+pub fn build_in_memory_engine(
```
Comment (SpadeA-Tang) [inline components/server/src/common.rs:698]: rename to build_hybrid_engine to make it aligh with return type?

File: components/raftstore/src/coprocessor/read_write/write_batch.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+pub trait WriteBatchObserver: Send {
+    fn create_observable_write_batch(&self) -> Box<dyn ObservableWriteBatch>;
+}
+
+/// A simplified version of `engine_trait::WriteBatch` that observe the write
+/// operations of a `engine_trait::WriteBatch`.
+///
+/// It exists because raftstore coprocessor only accepts trait object while
+/// the original `engine_trait::WriteBatch` can't be a trait object.
```
Comment (SpadeA-Tang) [inline components/raftstore/src/coprocessor/read_write/write_batch.rs:15]: why it cannot be a trait object?

File: components/hybrid_engine/src/observer/snapshot.rs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+use raftstore::coprocessor::{
+    dispatcher::BoxSnapshotObserver, CoprocessorHost, ObservedSnapshot, SnapshotObserver,
+};
+use range_cache_memory_engine::{RangeCacheMemoryEngine, RangeCacheSnapshot};
+
+use crate::new_in_memory_snapshot;
+
+/// RangeCacheSnapshotPin pins data of a RangeCacheMemoryEngine during taking
+/// snapshot. It prevents the data from being evicted or deleted from the cache.
+pub struct RangeCacheSnapshotPin {
```
Comment (glorv) [inline components/hybrid_engine/src/observer/snapshot.rs:13]: As discussed offline, this SnapshotObserver is useless and should be removed. If you don't want to do it, please at least add a `TODO`.

**Review by glorv — COMMENTED** [review]
Please add explaination about the overall architecture changes in the PR description.

File: components/hybrid_engine/src/observer/write_batch.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+use engine_traits::{is_data_cf, CacheRegion, KvEngine, Mutable, Result, WriteBatch, WriteOptions};
+use raftstore::coprocessor::{
+    dispatcher::BoxWriteBatchObserver, Coprocessor, CoprocessorHost, ObservableWriteBatch,
+    WriteBatchObserver,
+};
+use range_cache_memory_engine::{RangeCacheMemoryEngine, RangeCacheWriteBatch};
+
+#[derive(Clone)]
+pub struct HybridWriteBatchObserver {
```
Comment (SpadeA-Tang) [inline components/hybrid_engine/src/observer/write_batch.rs:11]: Under the current implementation, it reflects nothing about hybrid anymore. May give another name like RegionCacheWriteBatchObserver?

File: components/hybrid_engine/src/observer/write_batch.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+impl WriteBatchObserver for HybridWriteBatchObserver {
+    fn create_observable_write_batch(&self) -> Box<dyn ObservableWriteBatch> {
+        Box::new(HybridObservableWriteBatch {
+            cache_write_batch: RangeCacheWriteBatch::from(&self.cache_engine),
+        })
+    }
+}
+
+struct HybridObservableWriteBatch {
```
Comment (SpadeA-Tang) [inline components/hybrid_engine/src/observer/write_batch.rs:37]: ditto

File: components/server/src/server.rs [hunk 9]
```diff
@@ -1658,23 +1644,39 @@ where
...
             RangeCacheEngineContext::new(range_cache_engine_config.clone(), self.pd_client.clone());
         let range_cache_engine_statistics = range_cache_engine_context.statistics();
-        let kv_engine: EK = KvEngineBuilder::build(
-            range_cache_engine_context,
-            disk_engine.clone(),
-            Some(self.pd_client.clone()),
-            Some(Arc::new(self.region_info_accessor.clone())),
-        );
+        let in_memory_engine = if self.core.config.range_cache_engine.enabled {
+            let in_memory_engine = build_hybrid_engine(
```
Comment (SpadeA-Tang) [inline components/server/src/server.rs:1648]: Do you plan to remove hybrid enigne or rename it in the following PR? I think now this name is confusing.

File: components/raftstore/src/coprocessor/dispatcher.rs [hunk 10]
```diff
@@ -314,6 +326,8 @@ where
     update_safe_ts_observers: Vec<Entry<BoxUpdateSafeTsObserver>>,
     message_observers: Vec<Entry<BoxMessageObserver>>,
     region_heartbeat_observers: Vec<Entry<BoxRegionHeartbeatObserver>>,
+    write_batch_observer: Option<BoxWriteBatchObserver>,
```
[the remark this answers is not in the document]
Comment (overvenus) [reply components/raftstore/src/coprocessor/dispatcher.rs:329]: It's okay to change it to `Vec<BoxWriteBatchObserver>`, though write batch handling may be more tedious. I'll leave some notes here.

File: components/server/src/server.rs (hunk 9 above)
Comment (overvenus) [reply components/server/src/server.rs:1648]: Yes, will do clean up in https://github.com/tikv/tikv/pull/17380.

File: components/hybrid_engine/src/observer/snapshot.rs (hunk 6 above)
Comment (overvenus) [reply components/hybrid_engine/src/observer/snapshot.rs:15]: Added, and will open an issue for tracking the TODO.

Comment (overvenus) [reply components/hybrid_engine/src/observer/snapshot.rs:15]: Here is the issue https://github.com/tikv/tikv/issues/17492

**Review by SpadeA-Tang — APPROVED** [review]
